include/net/xia_ether.h

```diff
@@ -0,0 +1,347 @@
+
+	/* WARNING: @xhm_common is of variable size, and
+	 * MUST be the last member of the struct.
+	 */
+	struct fib_xid		xem_common;
+};
+
+struct xia_header_ops {
+	int	(*create) (struct sk_buff *skb, struct net_device *dev,
+			   unsigned short type, const void *daddr,
+			   const void *saddr, unsigned int len);
+	int	(*parse)(const struct sk_buff *skb, unsigned char *haddr);
+	int	(*cache)(const struct fib_xid_ether_main *mfxid, struct hh_cache *hh, __be16 type);
+	void	(*cache_update)(struct hh_cache *hh,
+				const struct net_device *dev,
+				const unsigned char *haddr,
+			    const int type);
+};
+
+int xia_ether_header_cache(const struct fib_xid_ether_main *mfxid, struct hh_cache *hh, __be16 type)
+{
+	struct ethhdr *eth;
+	const struct net_device *dev = mfxid->host_interface;
+
+	eth = (struct ethhdr *)
+	    (((u8 *) hh->hh_data) + (HH_DATA_OFF(sizeof(*eth))));
+
+	if (type != htons(ETH_P_XIP))
+		return -1;
+
+	eth->h_proto = type;
+	memcpy(eth->h_source, dev->dev_addr, ETH_ALEN);
+	memcpy(eth->h_dest, mfxid->neigh_addr->ha, ETH_ALEN);
+	hh->hh_len = ETH_HLEN;
+	return 0;
+}
+
+void xia_ether_header_cache_update(struct hh_cache *hh,
+			     const struct net_device *dev,
+			     const unsigned char *haddr,
+			     const int type)
+{
+	struct ethhdr *eth;
+	eth = (struct ethhdr *)
+		    (((u8 *) hh->hh_data) + (HH_DATA_OFF(sizeof(*eth))));
+	if(type)
+		memcpy(eth->h_dest , haddr, ETH_ALEN);
+	else
+		memcpy(eth->h_source, dev->dev_addr, ETH_ALEN);
+}
+
+const struct xia_header_ops xia_ether_hdr_ops ____cacheline_aligned = {
+	.create		= eth_header,
+	.parse		= eth_header_parse,
+	.cache		= xia_ether_header_cache,
+	.cache_update	= xia_ether_header_cache_update,
+};
+
+static void mfxid_update_hhs(struct fib_xid_ether_main *mfxid, const int type)
+{
+	struct hh_cache *hh;
+	hh = &mfxid->cached_hdr;
+
+	if (hh->hh_len) {
+		write_seqlock_bh(&hh->hh_lock);
+		xia_ether_hdr_ops.cache_update(hh, mfxid->host_interface, mfxid->neigh_addr->ha, type);
+		write_sequnlock_bh(&hh->hh_lock);
+	}
+}
+
+static inline void einterface_hold(struct ether_interface *eint)
+{
+	atomic_inc(&eint->refcnt);
+}
+
+static int mfxid_blackhole(struct fib_xid_ether_main *mfxid, struct sk_buff *skb)
+{
+	kfree_skb(skb);
+	return -ENETDOWN;
+}
+
+static void mfxid_hh_init(struct fib_xid_ether_main *mfxid)
+{
+	__be16 prot = htons(ETH_P_XIP);
+	struct hh_cache	*hh = &mfxid->cached_hdr;
+
+	write_lock_bh(&mfxid->chdr_lock);
+	/* Only one thread can come in here and initialize the
+	 * hh_cache entry and must prevent all the others from
+	 * accessing the same untill it is initialized.
+	 */
+	if (!hh->hh_len)
+		xia_ether_hdr_ops.cache(mfxid, hh, prot);
+
+	write_unlock_bh(&mfxid->chdr_lock);
+}
+
+static inline int cmp_addr(struct interface_addr *addr, const u8 *str_ha, struct net_device *dev)
+{
+	int c1 = memcmp(addr->ha, str_ha, dev->addr_len);
+	int c2 = addr->outgress_interface == dev;
+
+	if (likely(!c1 && c2)) {
```

Comment:
Remove the brace. Do not unnecessarily use braces where a single statement will do.